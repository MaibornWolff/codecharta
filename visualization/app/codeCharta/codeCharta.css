*, *:before, *:after {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
}

html, body {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    font-size: 1em;
    cursor: default;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#codeMap {
    position: absolute;
    z-index: 0;
    top: 50px;
    left: 0;
    width: 100%;
    height: 100%;
}

#hud {
    position: absolute;
    z-index: 10;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    max-width: none !important;
    pointer-events: none;
    padding-left:10px;
    padding-right:10px;
}

revision-chooser-component, legend-panel-component, file-chooser-panel-directive, settings-panel-component, .mwlink {
    pointer-events: auto;
}

#title {
    text-transform: uppercase;
    font-size: 0.9em;
    line-height: 1.4em;
    float: right;
    margin: 0.5em 1.6em 0 0;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#title:first-child {
    font-size: 1.82em;
}

#title:last-child {
    font-size: 1.5em;
}

.collapsible > li {
    background-color: #ffffff !important;
}

.col-backdrop {
    background-color: rgba(0,0,0,0.1);
}

.bottom-row {
    position: fixed;
    bottom: 0;
    width:100%;
}

.mwlink > img{
    width: 100%;
    max-width: 4em;
    float:right;
}

.chip {
    background-color: #efefef;
}

.row, .col {
    min-height: 0px !important;
}


/*

#actionButtons {
    position: fixed;
    bottom: 2em;
    right: 2em;
    z-index: 1000;
}

#settingsPanel {
    position: fixed;
    right: 0;
    width: 0px;
    height: 100%;
    background: #fff;
    z-index: 1000;
    box-shadow: 0 0 10px #000;
}

#detailWindow {
    position: fixed;
    z-index: 999;
    bottom: 0.5em;
    left: 0.5em;
    color:#5A585A;
    font-size: 0.8em;
}

#legendButton {
    position: absolute;
    top: 11.5em;
    left: -2.9em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#legendPanel {
    position: fixed;
    top: 9em;
    left: -500px;
}

#revisionButton {
    position: absolute;
    top: 21em;
    left: -3.5em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#mapButton {
    position: absolute;
    top: 21em;
    left: -2.3em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#revisionChooser {
    position: fixed;
    top: 18.5em;
    left: -500px;
}

.rzslider .rz-pointer, .rzslider .rz-selection {
    background: #26a69a !important;
}

.rzslider .rz-bar {
    background: rgb(196,196,196);
}*/