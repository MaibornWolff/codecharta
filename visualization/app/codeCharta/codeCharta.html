<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>