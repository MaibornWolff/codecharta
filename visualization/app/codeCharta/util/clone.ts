import { klona } from "klona/json"

export function clone<T>(content: T): T {
	return klona(content)
}
