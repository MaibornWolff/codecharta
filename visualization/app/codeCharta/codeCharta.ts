"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


