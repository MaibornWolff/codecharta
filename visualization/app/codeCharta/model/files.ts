import { CCFile, FileSelectionState, FileState } from "../codeCharta.model"
const clone = require("rfdc")()

export class Files {
	constructor(private files: FileState[] = []) {}

	public getFiles(): FileState[] {
		return this.files
	}

	public reset() {
		this.files = []
	}

	public resetSelection() {
		this.files = this.files.map(x => {
			return { ...x, selectedAs: FileSelectionState.None }
		})
	}

	public getCCFiles(): CCFile[] {
		return this.files.map(x => x.file)
	}

	public addFile(file: CCFile) {
		this.files = [...this.files, { file: clone(file), selectedAs: FileSelectionState.None }]
	}

	public setFiles(files: Files) {
		this.files = clone(files.getFiles())
	}

	public setSingleByName(fileName: string) {
		this.files = this.files.map(x => {
			if (x.file.fileMeta.fileName === fileName) {
				return { ...x, selectedAs: FileSelectionState.Single }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setSingle(file: CCFile) {
		this.setSingleByName(file.fileMeta.fileName)
	}

	public setDeltaByNames(referenceFileName: string, comparisonFileName: string) {
		this.files = this.files.map(x => {
			if (x.file.fileMeta.fileName === referenceFileName) {
				return { ...x, selectedAs: FileSelectionState.Reference }
			} else if (x.file.fileMeta.fileName === comparisonFileName) {
				return { ...x, selectedAs: FileSelectionState.Comparison }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.setDeltaByNames(reference.fileMeta.fileName, comparison.fileMeta.fileName)
	}

	public setMultipleByNames(partialFileNames: string[]) {
		this.files = this.files.map(x => {
			if (partialFileNames.includes(x.file.fileMeta.fileName)) {
				return { ...x, selectedAs: FileSelectionState.Partial }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.setMultipleByNames(multipleFiles.map(x => x.fileMeta.fileName))
	}

	public fileStatesAvailable(): boolean {
		return this.getVisibleFileStates().length > 0
	}

	public getFileByFileName(fileName: string): CCFile {
		const matchingFileState: FileState = this.files.find(x => x.file.fileMeta.fileName == fileName)
		return matchingFileState ? matchingFileState.file : undefined
	}

	public getVisibleFiles(): CCFile[] {
		return this.getVisibleFileStates().map(x => x.file)
	}

	public getVisibleFileStates(): FileState[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None)
	}

	public isSingleState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Single)
	}

	public isDeltaState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Reference || x.selectedAs === FileSelectionState.Comparison)
	}

	public isPartialState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Partial)
	}
}
