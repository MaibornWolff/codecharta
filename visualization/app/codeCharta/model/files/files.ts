import { CCFile } from "../../codeCharta.model"

export interface FileState {
	file: CCFile
	selectedAs: FileSelectionState
}

export enum FileSelectionState {
	Single = "Single",
	Reference = "Reference",
	Comparison = "Comparison",
	Partial = "Partial",
	None = "None"
}
