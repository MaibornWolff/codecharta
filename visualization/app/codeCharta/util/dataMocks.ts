import { CCFile, CodeMapNode, Edge, MetricData, Node, Settings } from "../codeCharta.model"
import { CodeMapBuilding } from "../ui/codeMap/rendering/codeMapBuilding"
import * as THREE from "three"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: {},
	blacklist: [],
	edges: [
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/big leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		},
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file_2018-12-14_9-39.cc.json",
	nodes: [
		{
			attributes: {},
			children: [
				{
					attributes: { Functions: 10, MCC: 1, RLOC: 100 },
					link: "http://www.google.de",
					name: "big leaf",
					type: "File"
				},
				{
					attributes: {},
					children: [
						{
							attributes: { Functions: 100, MCC: 100, RLOC: 30 },
							name: "small leaf",
							type: "File"
						},
						{
							attributes: { Functions: 1000, MCC: 10, RLOC: 70 },
							name: "other small leaf",
							type: "File"
						}
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: "/root",
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: new THREE.Vector3(1, 1.8, 1),
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}

export const DEFAULT_SETTINGS: Settings = {
	appSettings: {
		amountOfTopLabels: 1,
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		dynamicMargin: true,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		invertHeight: false,
		isWhiteBackground: false,
		mapColors: {
			angularGreen: "#00BFA5",
			base: "#666666",
			defaultC: "#89ACB4",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"],
			negative: "#820E0E",
			negativeDelta: "#ff0E0E",
			neutral: "#ddcc00",
			positive: "#69AE40",
			positiveDelta: "#69FF40",
			selected: "#EB8319"
		},
		maximizeDetailPanel: false,
		scaling: new THREE.Vector3(1, 1, 1),
		whiteColorBuildings: false
	},
	dynamicSettings: {
		areaMetric: null,
		colorMetric: null,
		focusedNodePath: "",
		heightMetric: null,
		margin: null,
		neutralColorRange: {
			flipped: false,
			from: null,
			to: null
		},
		searchPattern: "",
		searchedNodePaths: []
	},
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	treeMapSettings: { mapSize: 500 }
}

export const TEST_NODE_ROOT: Node = {
	name: "root",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 10,
	visible: true,
	path: "root",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0x000000",
	flat: false
}

export const TEST_NODE_LEAF: Node = {
	name: "root/big leaf",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 20,
	visible: true,
	path: "root/big leaf",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0xFFFFFF",
	flat: false
}

export const CODE_MAP_BUILDING: CodeMapBuilding = new CodeMapBuilding(
	1,
	new THREE.Box3(),
	TEST_NODE_ROOT,
	DEFAULT_SETTINGS.appSettings.mapColors.neutral
)

export const METRIC_DATA: MetricData[] = [
	{ name: "mcc", maxValue: 1, availableInVisibleMaps: true },
	{ name: "rloc", maxValue: 2, availableInVisibleMaps: true }
]
