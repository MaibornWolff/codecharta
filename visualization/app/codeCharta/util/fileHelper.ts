import { ExportBlacklistType, ExportCCFile, OldAttributeTypes } from "../codeCharta.api.model"
import { AttributeTypes, BlacklistItem, BlacklistType, CCFile } from "../codeCharta.model"

export function getCCFile(fileName: string, fileContent: ExportCCFile): CCFile {
	return {
		fileMeta: {
			fileName,
			fileChecksum: fileContent.fileChecksum,
			projectName: fileContent.projectName,
			apiVersion: fileContent.apiVersion
		},
		settings: {
			fileSettings: {
				edges: fileContent.edges || [],
				attributeTypes: getAttributeTypes(fileContent.attributeTypes),
				blacklist: potentiallyUpdateBlacklistTypes(fileContent.blacklist || []),
				markedPackages: fileContent.markedPackages || []
			}
		},
		map: fileContent.nodes[0]
	}
}

function getAttributeTypes(attributeTypes: AttributeTypes | OldAttributeTypes): AttributeTypes {
	if (!attributeTypes || Array.isArray(attributeTypes.nodes) || Array.isArray(attributeTypes.edges)) {
		return {
			nodes: {},
			edges: {}
		}
	}

	return {
		nodes: attributeTypes.nodes ?? {},
		edges: attributeTypes.edges ?? {}
	}
}

function potentiallyUpdateBlacklistTypes(blacklist): BlacklistItem[] {
	for (const entry of blacklist) {
		if (entry.type === ExportBlacklistType.hide) {
			entry.type = BlacklistType.flatten
		}
	}
	return blacklist
}
