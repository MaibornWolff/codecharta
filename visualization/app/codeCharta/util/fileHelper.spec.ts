import { ExportBlacklistType, ExportCCFile } from "../codeCharta.api.model"
import { AttributeTypeValue, BlacklistType, CCFile, NameDataPair } from "../codeCharta.model"
import { getCCFile, getSelectedFilesSize } from "./fileHelper"
import { TEST_FILE_CONTENT } from "./dataMocks"
import { clone } from "./clone"
import { FileSelectionState, FileState } from "../model/files/files"
import packageJson from "../../../package.json"

describe("FileHelper", () => {
	let fileContent: ExportCCFile

	beforeEach(() => {
		fileContent = clone(TEST_FILE_CONTENT)
	})

	describe("getCCFile", () => {
		it("should build a CCFile", () => {
			const nameDataPair: NameDataPair = { content: fileContent, fileName: "fileName", fileSize: 30 }
			const result = getCCFile(nameDataPair)

			expect(result).toMatchSnapshot()
		})

		it("should convert old blacklist type", () => {
			fileContent.blacklist = [{ path: "foo", type: ExportBlacklistType.hide }]

			const nameDataPair: NameDataPair = { content: fileContent, fileName: "fileName", fileSize: 30 }
			const result = getCCFile(nameDataPair)

			expect(result.settings.fileSettings.blacklist).toEqual([{ path: "foo", type: BlacklistType.flatten }])
		})

		it("should ignore old attribute types", () => {
			fileContent.attributeTypes = {
				nodes: [{ mcc: AttributeTypeValue.absolute }],
				edges: [{ pairingRate: AttributeTypeValue.relative }]
			}

			const nameDataPair: NameDataPair = { content: fileContent, fileName: "fileName", fileSize: 30 }
			const result = getCCFile(nameDataPair)

			expect(result.settings.fileSettings.attributeTypes).toEqual({ nodes: {}, edges: {} })
		})

		it("should return empty attributeTypes", () => {
			fileContent.attributeTypes = {}

			const nameDataPair: NameDataPair = { content: fileContent, fileName: "fileName", fileSize: 30 }
			const result = getCCFile(nameDataPair)

			expect(result.settings.fileSettings.attributeTypes).toEqual({ nodes: {}, edges: {} })
		})

		it("should return empty attributeTypes if the property doesn't exist", () => {
			fileContent.attributeTypes = undefined

			const nameDataPair: NameDataPair = { content: fileContent, fileName: "fileName", fileSize: 30 }
			const result = getCCFile(nameDataPair)

			expect(result.settings.fileSettings.attributeTypes).toEqual({ nodes: {}, edges: {} })
		})
	})

	describe("getSelectedFilesSize", () => {
		it("should sum up sizes of selected files", () => {
			const file1 = {
				fileMeta: {
					fileName: "file1",
					fileChecksum: "invalid-md5-checksum-1",
					projectName: "Sample Project",
					apiVersion: packageJson.codecharta.apiVersion,
					exportedFileSize: 10
				}
			} as CCFile

			const file2 = {
				fileMeta: {
					fileName: "file2",
					fileChecksum: "invalid-md5-checksum-2",
					projectName: "Sample Project",
					apiVersion: packageJson.codecharta.apiVersion,
					exportedFileSize: 10
				}
			} as CCFile

			const unselectedFile = {
				fileMeta: {
					fileName: "unselectedFile",
					fileChecksum: "invalid-md5-checksum-3",
					projectName: "Sample Project",
					apiVersion: packageJson.codecharta.apiVersion,
					exportedFileSize: 10
				}
			} as CCFile

			const severalFiles: FileState[] = [
				{
					file: file1,
					selectedAs: FileSelectionState.Partial
				},
				{
					file: file2,
					selectedAs: FileSelectionState.Comparison
				},
				{
					file: unselectedFile,
					selectedAs: FileSelectionState.None
				}
			]

			expect(getSelectedFilesSize(severalFiles)).toBe(20)
		})
	})
})
