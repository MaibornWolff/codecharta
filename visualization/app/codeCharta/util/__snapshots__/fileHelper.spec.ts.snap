// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileHelper getCCFile should build a CCFile 1`] = `
Object {
  "fileMeta": Object {
    "apiVersion": "1.2",
    "fileChecksum": "invalid-md5-sample",
    "fileName": "fileName",
    "projectName": "Sample Map",
  },
  "map": Object {
    "attributes": Object {},
    "children": Array [
      Object {
        "attributes": Object {
          "functions": 10,
          "mcc": 1,
          "rloc": 100,
        },
        "isExcluded": false,
        "isFlattened": false,
        "link": "http://www.google.de",
        "name": "big leaf",
        "type": "File",
      },
      Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 100,
              "mcc": 100,
              "rloc": 30,
            },
            "isExcluded": false,
            "isFlattened": false,
            "name": "small leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "isExcluded": false,
            "isFlattened": false,
            "name": "other small leaf",
            "type": "File",
          },
        ],
        "isExcluded": false,
        "isFlattened": false,
        "name": "Parent Leaf",
        "type": "Folder",
      },
    ],
    "isExcluded": false,
    "isFlattened": false,
    "name": "root",
    "type": "Folder",
  },
  "settings": Object {
    "fileSettings": Object {
      "attributeTypes": Object {
        "edges": Object {},
        "nodes": Object {},
      },
      "blacklist": Array [],
      "edges": Array [],
      "markedPackages": Array [],
    },
  },
}
`;
